solarforecastarbiter/metrics/calculator.py

```diff
@@ -507,3 +518,82 @@ def _transform_prob_forecast_value_and_prob(proc_fx_obs):
                                 name=col)
         fx_fx_prob.append((fx, fx_prob))
     return fx_fx_prob
+
+
+def calculate_event_metrics(proc_fx_obs, categories, metrics):
+    """
+    Calculate event metrics for the processed data using the provided
+    categories and metric types.
+
+    Parameters
+    ----------
+    proc_fx_obs : datamodel.ProcessedForecastObservation
+    categories : list of str
+        List of categories to compute metrics over.
+    metrics : list of str
+        List of metrics to be computed.
+
+    Returns
+    -------
+    solarforecastarbiter.datamodel.MetricResult
+        Contains all the computed metrics by categories.
+
+    Raises
+    ------
+    RuntimeError
+        If there is no forecast, observation timeseries data or no metrics
+        are specified.
+    """
+
+    out = {
+        'name': proc_fx_obs.name,
+        'forecast_id': proc_fx_obs.original.forecast.forecast_id,
+        'observation_id': proc_fx_obs.original.observation.observation_id
+    }
+
+    fx = proc_fx_obs.forecast_values
+    obs = proc_fx_obs.observation_values
+
+    # No data or metrics
+    if fx.empty:
+        raise RuntimeError("No Forecast timeseries data.")
+    elif obs.empty:
+        raise RuntimeError("No Observation timeseries data.")
+    elif len(metrics) == 0:
+        raise RuntimeError("No metrics specified.")
+
+    # Dataframe for grouping
+    df = pd.concat({'forecast': fx, 'observation': obs}, axis=1)
+
+    metric_vals = []
+    # Calculate metrics
+    for category in set(categories):
+        # total (special category)
+        if category == 'total':
+            index_category = lambda x: 0  # NOQA
+        else:
+            index_category = getattr(df.index, category)
+
+        # Calculate each metric
+        for metric_ in set(metrics):
+            # Group by category
+            for cat, group in df.groupby(index_category):
+
+                # Calculate
+                res = _apply_event_metric_func(
+                    metric_, group.forecast, group.observation
+                )
+
+                # Change category label of the group from numbers
+                # to e.g. January or Monday
+                if category == 'month':
+                    cat = calendar.month_abbr[cat]
+                elif category == 'weekday':
+                    cat = calendar.day_abbr[cat]
+
+                metric_vals.append(datamodel.MetricValue(
+                    category, metric_, str(cat), res))
+
+    out['values'] = tuple(metric_vals)
```

Comment:
https://github.com/SolarArbiter/solarforecastarbiter-core/pull/396 added a function to ensure that metrics are reported in a consistent order. This line should become

```
    out['values'] = _sort_metrics_vals(
        metric_vals, datamodel.ALLOWED_EVENT_METRICS)
```

copy/paste/adapt this test for a couple of event metrics - doesn't need to be super comprehensive

https://github.com/SolarArbiter/solarforecastarbiter-core/blob/0663659f97934764385a1fa471b3cb559e1ccc4f/solarforecastarbiter/metrics/tests/test_calculator.py#L224-L246

Reply:
Will do.

Reply:
Done.